Replace potentially more confusing decltype() usage with template lambdas where possible

weinig opened the pull request:
#### 5fe8dd6a3b46ad9e11f397ea324c5615cb7680bf
<pre>
Replace potentially more confusing decltype() usage with template lambdas where possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=273694">https://bugs.webkit.org/show_bug.cgi?id=273694</a>

Reviewed by Darin Adler.

A somewhat common patter in WebCore is for a call site to use a generic lambda but still
want to know what the type of the generic argument was. Something like:

    function([](auto parameter) {
        using Type = std::decay_t&lt;decltype(parameter)&gt;;

        ...
    });

This was necessary in the past due to the lack of template lambdas, but now we have them
and this can be come:

    function([]&lt;typename Type&gt;(Type parameter) {
        ...
    });

Additionally, this now lets us provide a template parameter without the need for passing
an argument. For example, if the actual value of the parameter was not needed in the
example above, we could instead write:

    function([]&lt;typename Type&gt;() {
        ...
    });

This allows us to generalize a few places that needed a mapping of color space to color
type, allowing us to remove some duplicate copies of the big switch.

* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::matchAll):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
(WebCore::WorkerCacheStorageConnection::retrieveRecords):
* Source/WebCore/Modules/entriesapi/FileSystemEntry.cpp:
(WebCore::FileSystemEntry::getParent):
* Source/WebCore/Modules/fetch/WindowOrWorkerGlobalScopeFetch.cpp:
(WebCore::doFetch):
* Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
(WebCore::IDBKeyData::IDBKeyData):
(WebCore::IDBKeyData::type const):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::addIceCandidate):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::addIceCandidate):
* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
(WebCore::Detail::forEachArgs):
(WebCore::forEach):
* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::appendColorInterpolationMethod):
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::equals const):
(WebCore::CSSValue::operator delete):
* Source/WebCore/css/DeprecatedCSSOMValue.cpp:
(WebCore::DeprecatedCSSOMValue::operator delete):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::operator delete):
(WebCore::StyleRuleBase::copy const):
* Source/WebCore/css/color/CSSResolvedColorMix.cpp:
(WebCore::mix):
* Source/WebCore/dom/Node.cpp:
(WebCore::NodeRareData::operator delete):
* Source/WebCore/loader/SubresourceIntegrity.cpp:
(WebCore::parseIntegrityMetadata):
* Source/WebCore/loader/appcache/ApplicationCacheManifestParser.cpp:
(WebCore::parseApplicationCacheManifest):
* Source/WebCore/platform/Length.cpp:
(WebCore::Length::Length):
* Source/WebCore/platform/graphics/Color.cpp:
(WebCore::Color::anyComponentIsNone const):
(WebCore::Color::invertedColorWithAlpha const):
* Source/WebCore/platform/graphics/ColorBlending.cpp:
(WebCore::requiresLegacyInterpolationRules):
* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::convertAndResolveColorComponents):
* Source/WebCore/platform/graphics/ColorInterpolation.cpp:
(WebCore::interpolateColors):
* Source/WebCore/platform/graphics/ColorInterpolationMethod.cpp:
(WebCore::serializationForCSS):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/ColorInterpolationMethod.h:
(WebCore::add):
(WebCore::serializationForCSS):
* Source/WebCore/platform/graphics/ColorSpace.h:
(WebCore::callWithColorType):
* Source/WebCore/platform/graphics/PathSegment.cpp:
(WebCore::PathSegment::applyElements const):
(WebCore::PathSegment::transform):
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
(WebCore::GradientRendererCG::makeShading const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::isValid):
(WebCore::DisplayList::dumpItem):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::lastLineBaseline const):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::calculateApproximateStrokeBoundingBox):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::RuleFeatureSet::collectFeatures):
* Source/WebCore/svg/SVGViewSpec.cpp:
(WebCore::SVGViewSpec::parseViewSpec):
* Source/WebCore/workers/service/ServiceWorkerClients.cpp:
(WebCore::ServiceWorkerClients::get):
(WebCore::ServiceWorkerClients::openWindow):
* Source/WebCore/workers/service/context/ServiceWorkerFetch.cpp:
(WebCore::ServiceWorkerFetch::dispatchFetchEvent):

Canonical link: <a href="https://commits.webkit.org/278358@main">https://commits.webkit.org/278358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34396ed8844c9ca8949a1b216936026c4506d5b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50233 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2530 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53492 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52536 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35753 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/548 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40988 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52332 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27196 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43237 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22086 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24616 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/489 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8611 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46603 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/551 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55077 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25329 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/484 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48400 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26590 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43424 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47423 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11034 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27454 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26322 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
This looks like an excellent change: Would love to add a note to the style guide about this principle

Commit:
Replace potentially more confusing decltype() usage with template lambdas where possible
https://bugs.webkit.org/show_bug.cgi?id=273694

Reviewed by Darin Adler.

A somewhat common patter in WebCore is for a call site to use a generic lambda but still
want to know what the type of the generic argument was. Something like:

    function([](auto parameter) {
        using Type = std::decay_t<decltype(parameter)>;

        ...
    });

This was necessary in the past due to the lack of template lambdas, but now we have them
and this can be come:

    function([]<typename Type>(Type parameter) {
        ...
    });

Additionally, this now lets us provide a template parameter without the need for passing
an argument. For example, if the actual value of the parameter was not needed in the
example above, we could instead write:

    function([]<typename Type>() {
        ...
    });

This allows us to generalize a few places that needed a mapping of color space to color
type, allowing us to remove some duplicate copies of the big switch.

* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::matchAll):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
(WebCore::WorkerCacheStorageConnection::retrieveRecords):
* Source/WebCore/Modules/entriesapi/FileSystemEntry.cpp:
(WebCore::FileSystemEntry::getParent):
* Source/WebCore/Modules/fetch/WindowOrWorkerGlobalScopeFetch.cpp:
(WebCore::doFetch):
* Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
(WebCore::IDBKeyData::IDBKeyData):
(WebCore::IDBKeyData::type const):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::addIceCandidate):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::addIceCandidate):
* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
(WebCore::Detail::forEachArgs):
(WebCore::forEach):
* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::appendColorInterpolationMethod):
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::equals const):
(WebCore::CSSValue::operator delete):
* Source/WebCore/css/DeprecatedCSSOMValue.cpp:
(WebCore::DeprecatedCSSOMValue::operator delete):
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::operator delete):
(WebCore::StyleRuleBase::copy const):
* Source/WebCore/css/color/CSSResolvedColorMix.cpp:
(WebCore::mix):
* Source/WebCore/dom/Node.cpp:
(WebCore::NodeRareData::operator delete):
* Source/WebCore/loader/SubresourceIntegrity.cpp:
(WebCore::parseIntegrityMetadata):
* Source/WebCore/loader/appcache/ApplicationCacheManifestParser.cpp:
(WebCore::parseApplicationCacheManifest):
* Source/WebCore/platform/Length.cpp:
(WebCore::Length::Length):
* Source/WebCore/platform/graphics/Color.cpp:
(WebCore::Color::anyComponentIsNone const):
(WebCore::Color::invertedColorWithAlpha const):
* Source/WebCore/platform/graphics/ColorBlending.cpp:
(WebCore::requiresLegacyInterpolationRules):
* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::convertAndResolveColorComponents):
* Source/WebCore/platform/graphics/ColorInterpolation.cpp:
(WebCore::interpolateColors):
* Source/WebCore/platform/graphics/ColorInterpolationMethod.cpp:
(WebCore::serializationForCSS):
(WebCore::operator<<):
* Source/WebCore/platform/graphics/ColorInterpolationMethod.h:
(WebCore::add):
(WebCore::serializationForCSS):
* Source/WebCore/platform/graphics/ColorSpace.h:
(WebCore::callWithColorType):
* Source/WebCore/platform/graphics/PathSegment.cpp:
(WebCore::PathSegment::applyElements const):
(WebCore::PathSegment::transform):
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
(WebCore::GradientRendererCG::makeShading const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::isValid):
(WebCore::DisplayList::dumpItem):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::lastLineBaseline const):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::calculateApproximateStrokeBoundingBox):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::RuleFeatureSet::collectFeatures):
* Source/WebCore/svg/SVGViewSpec.cpp:
(WebCore::SVGViewSpec::parseViewSpec):
* Source/WebCore/workers/service/ServiceWorkerClients.cpp:
(WebCore::ServiceWorkerClients::get):
(WebCore::ServiceWorkerClients::openWindow):
* Source/WebCore/workers/service/context/ServiceWorkerFetch.cpp:
(WebCore::ServiceWorkerFetch::dispatchFetchEvent):

Canonical link: https://commits.webkit.org/278358@main
